Fix issue in Freebsd when building with GCC 14.1.0

bjosv opened the pull request:
_Note: This is a PR directly to branch `1.21`._

Since GCC 14 now [implements](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60512) `__has_extension()` the build fails while building the F-stack `lib`.
The reason for the failure is the use of Clangs atomics APIs (`__c11_*`) which are only provided by Clang.

Fix the build to use C11 atomics when available in GCC by incorporating a fix already included in latest FreeBSD [stable/12](https://github.com/freebsd/freebsd-src/blame/stable/12/sys/sys/stdatomic.h#L174).
See https://reviews.freebsd.org/D16585 which enables C11 atomics when building with GCC >= 4.7.

Previously the non-native support for _Atomic() was used, placing objects in a structure to prevent direct non-atomic access,
but `_Atomic(T)` did not get defined for GCC-14.

**For reference**
See section `C family` regarding atomics in GCC-4.7 and has_extension() in GCC-14:
https://gcc.gnu.org/gcc-4.7/changes.html
https://gcc.gnu.org/gcc-14/changes.html

bjosv committed:
Fix issue in Freebsd when building with GCC 14.1.0

Since GCC 14 now implements `__has_extension()` the build started to use
Clang atomics, i.e. the `__c11_*` APIs only provided by Clang.

Fix build to use GNUC atomics and include the upstream patch:
  https://reviews.freebsd.org/D16585
which enables C11 atomics when building with GCC >= 4.7.

jfb8856606 commented:
Redhat 7.x/CentOS 7.x used gcc-4.8.5, it > 4.7, but it build failed.

jfb8856606 commented:
> GCC began to offer full support for the C11 standard starting with version 4.9, including support for the `_Atomic` keyword. This means that if you want to use atomic operations and types as defined in the C11 standard, you should use GCC version 4.9 or later.
-- from GPT4

And see: [https://gcc.gnu.org/gcc-4.9/changes.html](https://gcc.gnu.org/gcc-4.9/changes.html)

jfb8856606 commented on freebsd/sys/cdefs.h:
```diff
@@ -276,7 +276,7 @@
 
 #if !defined(__cplusplus) && !__has_extension(c_atomic) && \
-    !__has_extension(cxx_atomic)
+    !__has_extension(cxx_atomic) && !__GNUC_PREREQ__(4, 7)
```
Shoud use `&& !__GNUC_PREREQ__(4, 9)`, see https://gcc.gnu.org/gcc-4.9/changes.html

bjosv replied:
Great finding, seems correct that 4.9 is the boundary

bjosv commented on freebsd/sys/stdatomic.h:
```diff
@@ -33,7 +33,7 @@
-#if __has_extension(c_atomic) || __has_extension(cxx_atomic)
+#if defined(__clang__) && (__has_extension(c_atomic) || __has_extension(cxx_atomic))
 #define	__CLANG_ATOMICS
 #elif __GNUC_PREREQ__(4, 7)
```
ah, interesting. I guess we need to update here then as well?

bjosv committed:
Use GNUC atomics from GCC 4.9

GCC began to offer full support for the C11 standard starting with version 4.9,
including support for the _Atomic keyword.

https://gcc.gnu.org/gcc-4.9/changes.html

bjosv commented:
Thanks, updated.
Should we add a Github CI to just build with different compilers? I can prepare a PR if you believe it would be beneficial.

jfb8856606 commented:
PRs for CI are very welcome.

There was CI in the early stage, but it lacked necessary updates and maintenance in the later stage.

bjosv commented:
Great, I'll take a look at it then.
I believe the previously used travis-ci have moved away from their initial free tier, so Github Actions is probably the easiest way forward.
